0.12 Nested ConfigMode: schema.SchemaConfigModeAttr Unexpectedly Requiring Attribute Syntax

### Terraform Version

```
# go.mod (master as of 2019-04-03)
github.com/hashicorp/terraform v0.12.0-alpha4.0.20190403225717-4d52999538ba
```

### Terraform Schema and Relevant Configuration

```go
"step": {
	Type:       schema.TypeList,
	Optional:   true,
	Computed:   true,
	ForceNew:   true,
	ConfigMode: schema.SchemaConfigModeAttr,
	Elem: &schema.Resource{
		Schema: map[string]*schema.Schema{
			"action_on_failure": {
				Type:     schema.TypeString,
				Required: true,
				ForceNew: true,
				ValidateFunc: validation.StringInSlice([]string{
					emr.ActionOnFailureCancelAndWait,
					emr.ActionOnFailureContinue,
					emr.ActionOnFailureTerminateCluster,
					emr.ActionOnFailureTerminateJobFlow,
				}, false),
			},
			"hadoop_jar_step": {
				Type:       schema.TypeList,
				MaxItems:   1,
				Required:   true,
				ForceNew:   true,
				ConfigMode: schema.SchemaConfigModeAttr,
				Elem: &schema.Resource{
					Schema: map[string]*schema.Schema{
						"args": {
							Type:     schema.TypeList,
							Optional: true,
							ForceNew: true,
							Elem:     &schema.Schema{Type: schema.TypeString},
						},
						"jar": {
							Type:     schema.TypeString,
							Required: true,
							ForceNew: true,
						},
						"main_class": {
							Type:     schema.TypeString,
							Optional: true,
							ForceNew: true,
						},
						"properties": {
							Type:     schema.TypeMap,
							Optional: true,
							ForceNew: true,
						},
					},
				},
			},
			"name": {
				Type:     schema.TypeString,
				Required: true,
				ForceNew: true,
			},
		},
	},
},
```

```hcl
resource "aws_emr_cluster" "tf-test-cluster" {
  # ... other configuration ...
  
  step {
    action_on_failure = "TERMINATE_CLUSTER"
    name              = "Setup Hadoop Debugging"

    hadoop_jar_step {
      jar  = "command-runner.jar"
      args = ["state-pusher-script"]
    }
  }
}
```

### Expected Behavior

Terraform 0.12 should allow configuration block syntax of children configuration blocks when the parent configuration block attribute has `ConfigMode: schema.SchemaConfigModeAttr`

### Actual Behavior

```
config is invalid: Unsupported block type: Blocks of type "hadoop_jar_step" are not expected here. Did you mean to define argument "hadoop_jar_step"? If so, use the equals sign to assign it a value.
```

### Steps to Reproduce

In the Terraform AWS Provider:

* Ensure the provider is updated to a very recent Terraform 0.12 Provider SDK, e.g.:

```console
$ go get github.com/hashicorp/terraform@4d52999538ba5b7ce9c1398b42b81d14ec3d895e
$ go mod tidy
$ go mod vendor
```

* Add `ConfigMode: schema.SchemaConfigModeAttr,` to the `aws_emr_cluster` resource `step` and `hadoop_jar_step` schema attributes

* Add the new acceptance test and two test configuration functions:

```go
func TestAccAWSEMRCluster_Step_ConfigMode(t *testing.T) {
	var cluster1, cluster2, cluster3 emr.Cluster
	rInt := acctest.RandInt()
	resourceName := "aws_emr_cluster.tf-test-cluster"

	resource.ParallelTest(t, resource.TestCase{
		PreCheck:     func() { testAccPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccCheckAWSEmrDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccAWSEmrClusterConfig_Step_Single(rInt),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSEmrClusterExists(resourceName, &cluster1),
					resource.TestCheckResourceAttr(resourceName, "step.#", "1"),
				),
			},
			{
				ResourceName:            resourceName,
				ImportState:             true,
				ImportStateVerify:       true,
				ImportStateVerifyIgnore: []string{"cluster_state", "configurations", "keep_job_flow_alive_when_no_steps"},
			},
			{
				Config: testAccAWSEmrClusterConfig_Step_NoBlocks(rInt),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSEmrClusterExists(resourceName, &cluster2),
					resource.TestCheckResourceAttr(resourceName, "step.#", "1"),
				),
			},
			{
				ResourceName:            resourceName,
				ImportState:             true,
				ImportStateVerify:       true,
				ImportStateVerifyIgnore: []string{"cluster_state", "configurations", "keep_job_flow_alive_when_no_steps"},
			},
			{
				Config: testAccAWSEmrClusterConfig_Step_Zeroed(rInt),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckAWSEmrClusterExists(resourceName, &cluster3),
					resource.TestCheckResourceAttr(resourceName, "step.#", "0"),
				),
			},
			{
				ResourceName:            resourceName,
				ImportState:             true,
				ImportStateVerify:       true,
				ImportStateVerifyIgnore: []string{"cluster_state", "configurations", "keep_job_flow_alive_when_no_steps"},
			},
		},
	})
}

func testAccAWSEmrClusterConfig_Step_NoBlocks(rInt int) string {
	return testAccAWSEmrClusterConfig_Step(rInt, "")
}

func testAccAWSEmrClusterConfig_Step_Zeroed(rInt int) string {
	return testAccAWSEmrClusterConfig_Step(rInt, "step = []")
}
```

* Run the acceptance testing

```console
$ TF_ACC=1 go test ./aws -v -timeout 120m -parallel 20 -run='TestAccAWSEMRCluster_Step_ConfigMode'
...
--- FAIL: TestAccAWSEMRCluster_Step_ConfigMode (0.82s)
    testing.go:568: Step 0 error: config is invalid: Unsupported block type: Blocks of type "hadoop_jar_step" are not expected here. Did you mean to define argument "hadoop_jar_step"? If so, use the equals sign to assign it a value.
```

### References

* https://github.com/hashicorp/terraform/issues/20505

--- a/lang/blocktoattr/fixup.go
+++ b/lang/blocktoattr/fixup.go
@@ -152,7 +152,8 @@ func (e *fixupBlocksExpr) Value(ctx *hcl.EvalContext) (cty.Value, hcl.Diagnostic
 	vals := make([]cty.Value, len(e.blocks))
 	var diags hcl.Diagnostics
 	for i, block := range e.blocks {
-		val, blockDiags := hcldec.Decode(block.Body, spec, ctx)
+		body := FixUpBlockAttrs(block.Body, schema)
+		val, blockDiags := hcldec.Decode(body, spec, ctx)
 		diags = append(diags, blockDiags...)
 		if val == cty.NilVal {
 			val = cty.UnknownVal(e.ety)

--- a/lang/blocktoattr/fixup_test.go
+++ b/lang/blocktoattr/fixup_test.go
@@ -216,7 +216,7 @@ foo {
 				}),
 			}),
 		},
-		"nested fixup": {
+		"fixup inside block": {
 			src: `
 container {
   foo {
@@ -237,6 +237,56 @@ container {
 					"container": {
 						Nesting: configschema.NestingList,
 						Block:   *fooSchema,
+					},
+				},
+			},
+			want: cty.ObjectVal(map[string]cty.Value{
+				"container": cty.ListVal([]cty.Value{
+					cty.ObjectVal(map[string]cty.Value{
+						"foo": cty.ListVal([]cty.Value{
+							cty.ObjectVal(map[string]cty.Value{
+								"bar": cty.StringVal("baz"),
+							}),
+							cty.ObjectVal(map[string]cty.Value{
+								"bar": cty.StringVal("boop"),
+							}),
+						}),
+					}),
+					cty.ObjectVal(map[string]cty.Value{
+						"foo": cty.ListVal([]cty.Value{
+							cty.ObjectVal(map[string]cty.Value{
+								"bar": cty.StringVal("beep value"),
+							}),
+						}),
+					}),
+				}),
+			}),
+		},
+		"fixup inside attribute-as-block": {
+			src: `
+container {
+  foo {
+    bar = "baz"
+  }
+  foo {
+    bar = "boop"
+  }
+}
+container {
+  foo {
+    bar = beep
+  }
+}
+`,
+			schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"container": {
+						Type: cty.List(cty.Object(map[string]cty.Type{
+							"foo": cty.List(cty.Object(map[string]cty.Type{
+								"bar": cty.String,
+							})),
+						})),
+						Optional: true,
 					},
 				},
 			},

--- a/lang/blocktoattr/variables_test.go
+++ b/lang/blocktoattr/variables_test.go
@@ -77,6 +77,39 @@ foo {
 				},
 			},
 		},
+		"block syntax with nested blocks": {
+			src: `
+foo {
+  bar {
+    boop = baz
+  }
+}
+`,
+			schema: &configschema.Block{
+				Attributes: map[string]*configschema.Attribute{
+					"foo": {
+						Type: cty.List(cty.Object(map[string]cty.Type{
+							"bar": cty.List(cty.Object(map[string]cty.Type{
+								"boop": cty.String,
+							})),
+						})),
+						Optional: true,
+					},
+				},
+			},
+			want: []hcl.Traversal{
+				{
+					hcl.TraverseRoot{
+						Name: "baz",
+						SrcRange: hcl.Range{
+							Filename: "test.tf",
+							Start:    hcl.Pos{Line: 4, Column: 12, Byte: 26},
+							End:      hcl.Pos{Line: 4, Column: 15, Byte: 29},
+						},
+					},
+				},
+			},
+		},
 		"dynamic block syntax": {
 			src: `
 dynamic "foo" {

